### PR TITLE
Compilation guide (addresses #11466)

### PR DESCRIPTION
**Pull request by drbig**
I've written a general Linux guide to give an overview of what is needed and also some explanation about the `make` flags.

I've also merged in the OS X guide and finally made the formatting consistent.

I believe that distro/OS specific guides should all live in a single file, as pretty much everything except for exact package names applies across all platforms.

### CODE REVIEW

File: COMPILING.md [hunk 1]
```diff
@@ -1,85 +1,319 @@
...
+
+This means you'll need a newer version of GCC (`g++`).
+
+The general rule is the newer the compiler the better.
+
+## Tools
+
+Most distros seem to package essential build tools as either a single package (Debian and derivatives have `build-essential`) or a package group (Arch has `base-devel`). You should use the above if available. Otherwise you'll at least need `make` and figure out the missing dependencies as you go.
+
+You should also install `git` and `gettext` (optional if you don't care about localizations).
```
Comment (narc0tiq) [inline COMPILING.md]: I'd rephrase this (as it implies git is optional if you don't care about localizations) -- perhaps include git in the previous line:

> otherwise you'll at least need `make`, and probably `git`, and figure out [...]

This would let us rephrase this line as:

> You may also wish to install `gettext` for localization support, but it is completely optional and you can compile without it.

Another notable mention would be Lua: you can technically compile without it, but you're not getting the full game if you do.

File: COMPILING.md [hunk 2]
```diff
@@ -1,85 +1,319 @@
...
+  * Curses: `ncurses`
+  * Tiles: `sdl2`, `sdl2_image`, `sdl2_ttf`, `sdl2_mixer`, `freetype2`
+
+E.g. for curses build on Debian and derivatives you'll also need `libncurses5-dev`.
+
+## Make flags
+
+Given you're building from source you have a number of choices to make:
+
+  * `NATIVE=` - you should use `linux64` if you're on a 64 bit system, `linux32` for 32 bit systems, and leave it out if you're building on Cygwin
```
Comment (narc0tiq) [inline COMPILING.md]: Uh, you should be able to leave it out on Linuxen unless you're explicitly cross-compiling (e.g. 32-to-64 or 64-to-32) -- I certainly don't include any `NATIVE=anything` in the jenkins builds.

File: COMPILING.md [hunk 3]
```diff
@@ -1,85 +1,319 @@
...
+  * `RELEASE=1` - without this you'll get a debug build (see note below)
+  * `TILES=1` - if you want the tiles version; note that with this the ncurses version won't be build
```
Comment (narc0tiq) [inline COMPILING.md]: Typo: "without this"

Comment (narc0tiq) [inline COMPILING.md]: Rephrase to something like "leave this out to build the ncurses version".

File: COMPILING.md [hunk 4]
```diff
@@ -1,85 +1,319 @@
...
+  * `SOUND=1` - if you want sound; this requires `TILES=1`
+  * `LOCALIZE=0` - this disables localizations so `gettext` is not needed
+  * `LUA=1` - this enables Lua *debug* support, may be useful if you plan writing mods
```
Comment (narc0tiq) [inline COMPILING.md]: Is that true? I thought it would omit all Lua support. Does it still link to liblua if you leave this out?

File: COMPILING.md [hunk 5]
```diff
@@ -1,85 +1,319 @@
...
-Cataclysm:DDA source directory.
+To build Cataclysm from source you will need at least a C++ compiler, some basic developer tools, and necessary build dependencies. The exact package names vary greatly from distro to distro, so this part of the guide is intended to give you higher-level understanding of the process.
+
+## Compiler
+
+You have three major choices here: GCC, Clang and MXE.
+
+  * GCC is almost always the default on Linux systems so it's likely you already have it
+  * Clang is usually faster than GCC, so it's worth installing if you plan to keep up with the latest experimentals
+  * MXE is a cross-compiler, so of any importance only if you plan to compile for Windows on your Linux machine (not covered here)
```
Comment (narc0tiq) [inline COMPILING.md]: I read this as saying MXE builds aren't covered by the document, so imagine my surprise when I found MXE compile instructions further below.

File: COMPILING.md [hunk 6]
```diff
@@ -1,85 +1,319 @@
...
-```
-
-### Building ###
-```
-PATH="${PATH}:~/src/mxe/usr/bin"
-make CROSS=i686-pc-mingw32-
-```
-
-## Linux (native) SDL builds ##
+    sudo apt-get install autoconf bison flex cmake git automake intltool libtool scons yasm
```
Comment (narc0tiq) [inline COMPILING.md]: Looks like you lost some backticks here -- this renders as HTML text, not preformatted.

**Comment by narc0tiq** [comment]
Looks pretty good, aside from the small issues I mentioned above. Admittedly, I only skimmed the OSX instructions, but they looked fine. 

Thanks for improving the documentation!

File: COMPILING.md (hunk 2 above)
Comment (kevingranade) [inline COMPILING.md]: An example of somewhere to use NATIVE* is building using distcc on a mixed
group of 32 and 64 bit systems.

File: COMPILING.md (hunk 4 above)
Comment (kevingranade) [inline COMPILING.md]: LUA is used in mods that adjust monster stats, faster and slower zombies
mods for example.

**Comment by drbig** [comment]
The Debian and OS X guides are verbatim as they were before sans the minor formatting changes.

I'm under impression that `lua` is an optional dependency... If I'm mistaken please correct me and I'll move it to general dependencies (I have lua installed anyway as a dependency for other stuff...).

**Comment by vickio** [comment]
There's an important change that should to be made to the OS X instructions since there is now a Homebrew formula for cataclysm that will automatically install the proper dependencies, but I can wait until after this gets merged to update the Mac stuff. I agree it should all be in one file.

**Comment by drbig** [comment]
This should be ready for merge as far as I can tell.

I'm thinking about writing a section on 'Keeping your CDDA up to date' that would be mostly about using git, ccache, doing stack traces and GitHub reporting (so the benefits go both ways :)) - but that can wait after the actual compilation instructions get updated.

**Comment by KA101** [comment]
Rare case where the makefile is changed yet it's clear for webmerge, since that was only a comment.

**Comment by drbig** [comment]
@vickio Feeling like updating the OS X part? I think it'll be worthwhile.
